cpp/src/lists/interleave_columns.cu

```diff
@@ -0,0 +1,359 @@
+
+    // Cannot call `cudf::strings::detail::make_strings_children` because that function executes the
+    // functor `comp_fn` on the range equal to `num_output_entries`.
+    auto [offsets_column, chars_column, null_mask, null_count] =
+      cudf::strings::detail::make_strings_children_with_null_mask(
+        comp_fn, num_output_lists, num_output_entries, stream, mr);
+
+    if (has_null_mask) {
+      return make_strings_column(num_output_entries,
+                                 std::move(offsets_column),
+                                 std::move(chars_column),
+                                 null_count,
+                                 std::move(null_mask),
+                                 stream,
+                                 mr);
+    }
+    return make_strings_column(num_output_entries,
+                               std::move(offsets_column),
+                               std::move(chars_column),
+                               0,
+                               rmm::device_buffer{},
+                               stream,
+                               mr);
+  }
+
+  template <class T>
+  std::enable_if_t<cudf::is_fixed_width<T>(), std::unique_ptr<column>> operator()(
+    table_view const& input,
+    column_view const& output_list_offsets,
+    size_type num_output_lists,
+    size_type num_output_entries,
+    bool has_null_mask,
+    rmm::cuda_stream_view stream,
+    rmm::mr::device_memory_resource* mr) const noexcept
+  {
+    auto const num_cols     = input.num_columns();
+    auto const num_rows     = input.num_rows();
+    auto const table_dv_ptr = table_device_view::create(input);
+
+    // The output child column.
+    auto const child_col = lists_column_view(*input.begin()).child();
+    auto output =
+      allocate_like(child_col, num_output_entries, mask_allocation_policy::NEVER, stream, mr);
+    auto output_dv_ptr = mutable_column_device_view::create(*output);
+
+    // The array of int8_t to store entry validities.
+    auto validities = rmm::device_uvector<int8_t>(has_null_mask ? num_output_entries : 0, stream);
+
+    thrust::for_each_n(
+      rmm::exec_policy(stream),
+      thrust::make_counting_iterator<size_type>(0),
+      num_output_lists,
+      [num_cols,
+       table_dv         = *table_dv_ptr,
+       d_validities     = validities.begin(),
+       dst_list_offsets = output_list_offsets.begin<offset_type>(),
+       d_output         = output_dv_ptr->begin<T>(),
+       has_null_mask] __device__(size_type const dst_list_id) {
+        auto const src_col_id     = dst_list_id % num_cols;
+        auto const src_list_id    = dst_list_id / num_cols;
+        auto const& src_lists_col = table_dv.column(src_col_id);
+        auto const src_list_offsets =
+          src_lists_col.child(lists_column_view::offsets_column_index).data<size_type>() +
+          src_lists_col.offset();
+        auto const& src_child = src_lists_col.child(lists_column_view::child_column_index);
```

Comment:
Same comment as above.  Maybe just "data" here.  Something to clearly distinguish it from the source column that contains it.